### PR TITLE
Validate subject mimetypes

### PR DESCRIPTION
**Pull request by srallen**
_Please request review from `@zooniverse/frontend` team. If PR is related to design, please request review from `@beckyrother` in addition._ 

Package: lib-classifier

Describe your changes:
Adds a mime type validation helper function that I've incorporated into the `createLocationCounts` helper, but can be used on its own too.

I took the approach that if any one of the locations of a subject is invalid to return an empty object with no counts. This isn't entirely incorporated into the `viewer` getter yet; it's not validating the viewers that explicitly look for a workflow configuration, but I wanted to get feedback first before doing any more work.

I also took a slightly different approach from the [original PFE version](https://github.com/zooniverse/Panoptes-Front-End/blob/ec86d183e6d1a1cce27b63ac51e338329ce40131/app/lib/get-subject-locations.js) using regexes and the test method. 

# Review Checklist

## General

- [ ] Are the tests passing locally and on Travis?
- [ ] Is the documentation up to date?

## Components
- [ ] Has a storybook story been created or updated?
- [ ] Is the component accessible? 
  - [ ] Can it be used with a screen reader? [BBC guide to testing with VoiceOver](https://bbc.github.io/accessibility-news-and-you/accessibility-and-testing-with-voiceover-os.html)
  - [ ] Can it be used from the keyboard? [WebAIM guide to keyboard testing](https://webaim.org/techniques/keyboard/#testing)
  - [ ] Is it passing accessibility checks in the storybook?

## Apps

- [ ] Does it work in all major browsers: Firefox, Chrome, Edge, Safari?
- [ ] Does it work on mobile?
- [ ] Can you `yarn panic && yarn bootstrap` or `docker-compose up --build` and app works as expected?

## Publishing

- [ ] Is the changelog updated?
- [ ] Are the dependencies updated for apps and libraries that are using the newly published library?

## Post-merging

- [ ] Did the app deploy to https://frontend.preview.zooniverse.org/projects/:project-name/:owner or https://frontend.preview.zooniverse.org/about?
- [ ] Is the new feature working or bug now fixed?
  - [ ] Is there a Talk or blog post written to announce the new feature(s)?
- [ ] Is the design working across browsers (Firefox, Chrome, Edge, Safari) and mobile?
  - [ ] Is this approved by our designer?
- [ ] Is this ready for production deployment?

### CODE REVIEW

File: packages/lib-classifier/src/helpers/validateMimeType/validateMimeType.spec.js [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+  'video/mp4',
+  'video/mpeg',
+  'video/x-m4v'
+]
+
+const invalidMimeTypes = [
+  'application/xml',
+  'audio/aac',
+  'audio/wave',
+  'image/tiff'
```
Comment (eatyourgreens) [inline packages/lib-classifier/src/helpers/validateMimeType/validateMimeType.spec.js]: Maybe add `text/html` to the list of blocked types?

**Comment by eatyourgreens** [comment]
I think the JSON data for the PH-TESS subjects might be marked as `text/plain` in the subjects served by Panoptes. It shows up in Talk using the text file viewer.

**Comment by eatyourgreens** [comment]
I think this will work very well with the `SingleImageSubject` model in #1557.

File: packages/lib-classifier/src/helpers/validateMimeType/validateMimeType.spec.js (hunk 1 above)
Comment (eatyourgreens) [reply packages/lib-classifier/src/helpers/validateMimeType/validateMimeType.spec.js]: I’m being paranoid, but could you block JS here too?

File: packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/SingleImageViewer/SingleImageViewerContainer.js [hunk 2]
```diff
@@ -69,7 +69,6 @@ class SingleImageViewerContainer extends React.Component {
   async preload () {
     const { subject } = this.props
     if (subject && subject.locations) {
-      // TODO: Validate for allowed image media mime types
       const imageUrl = Object.values(subject.locations[0])[0]
```
Comment (eatyourgreens) [inline packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/SingleImageViewer/SingleImageViewerContainer.js]: Where is this URL being validated?

File: packages/lib-classifier/src/helpers/createLocationCounts/createLocationCounts.js [hunk 3]
```diff
@@ -1,14 +1,26 @@
...
 
-  const images = types
-    .filter(type => type.startsWith('image'))
-    .length
+  try {
+    const images = mimeTypes
+      .filter(mimeType => mimeType.startsWith('image') && validateMimeType(mimeType))
+      .length
+
+    const json = mimeTypes
```
Comment (eatyourgreens) [inline packages/lib-classifier/src/helpers/createLocationCounts/createLocationCounts.js]: You often see `.js` and `.json` sent as `text/plain` on the web. I think PH-TESS subjects use `text/plain` for JSON data.

File: packages/lib-classifier/src/helpers/createLocationCounts/createLocationCounts.spec.js [hunk 4]
```diff
@@ -15,29 +15,53 @@ const mockLocations3 = [
...
+    it('should return the number of json files', function () {
+      const result4 = createLocationCounts(mockSubject(mockLocations4))
+
+      expect(result4.json).to.equal(1)
+    })
+  })
 
-    expect(result2.images).to.equal(2)
-    expect(result3.images).to.equal(2)
+  describe('when the mime types are invalid', function () {
```
Comment (eatyourgreens) [inline packages/lib-classifier/src/helpers/createLocationCounts/createLocationCounts.spec.js]: Is it worth having a text for a mix of valid and invalid types?

File: packages/lib-classifier/src/helpers/createLocationCounts/createLocationCounts.js (hunk 3 above)
Comment (srallen) [reply packages/lib-classifier/src/helpers/createLocationCounts/createLocationCounts.js]: I think TESS using that was due to a bug in the python client.

File: packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/SingleImageViewer/SingleImageViewerContainer.js (hunk 2 above)
Comment (srallen) [reply packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/SingleImageViewer/SingleImageViewerContainer.js]: `createLocationCounts` now uses the validation and if any of the locations are invalid, an empty object is returned. There's a conditional in the getter in the `Subject` store that will now evaluate as false and the viewer will return null. Then  in the `SubjectViewer` component, the return JSX component will be null, so this code will never run. I can add a test in that component to be sure.

File: packages/lib-classifier/src/helpers/createLocationCounts/createLocationCounts.js (hunk 3 above)
Comment (srallen) [reply packages/lib-classifier/src/helpers/createLocationCounts/createLocationCounts.js]: I didn't touch the getter that sets the viewer to use the LCV subject viewer to use the validation yet, so this is setup for VSLCV to validate properly and we can speak to the TESS team to make sure future sectors get uploaded with the correct mime type.

File: packages/lib-classifier/src/helpers/createLocationCounts/createLocationCounts.spec.js (hunk 4 above)
Comment (srallen) [reply packages/lib-classifier/src/helpers/createLocationCounts/createLocationCounts.spec.js]: I've added a test for this.

File: packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/SingleImageViewer/SingleImageViewerContainer.js (hunk 2 above)
Comment (srallen) [reply packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/SingleImageViewer/SingleImageViewerContainer.js]: I've added a couple of tests in `Subject.spec.js` for when the subject is invalid, the viewer should be null.

File: packages/lib-classifier/src/helpers/validateMimeType/validateMimeType.spec.js (hunk 1 above)
Comment (srallen) [reply packages/lib-classifier/src/helpers/validateMimeType/validateMimeType.spec.js]: I've added a few more invalid mime types.

**Comment by srallen** [comment]
@eatyourgreens This is ready for another review. I've added some more robust tests and another helper `validateSubjectLocations` which checks the mime type and also validates the file extension on the url. I'm still not totally certain how best to integrate this with the viewers that are using workflow config directly, so I'd like to have this reviewed as is and continue adding more checks in follow PRs if that's alright.

File: packages/lib-classifier/src/helpers/createLocationCounts/createLocationCounts.js (hunk 3 above)
Comment (eatyourgreens) [reply packages/lib-classifier/src/helpers/createLocationCounts/createLocationCounts.js]: Sounds good. You may find that the API client uploads `.json` as text, depending on where it's run and what type that machine/OS assigns to JSON files, but we can deal with that when it happens.

**Pull request merged**